### PR TITLE
lbry: init at 0.50.2

### LINKED ISSUE #78766: Package request: LBRY
**Project description**
Desktop client for the LBRY protocol. A YouTube alternative that uses the blockchain.

**Metadata**

* homepage URL: https://lbry.com/
* source URL: https://github.com/lbryio/lbry-desktop
* license: MIT
* platforms: Linux

### PR DESCRIPTION
**Pull request by enderger**
###### Motivation for this change
Closes #78766 and a YouTube comment conversation with DistroTube (one of the 2 packages he stated were missing from NixPkgs, the other being LibreWolf).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - × macOS (will not work due to ElectronJS using a completely different file structure on OSX, and I don't have any means to try to test a Mac builder. This would probably entail a separate PR)
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) (not entirely sure what this means)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/video/lbry/default.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    # desktop file/icon
+    mkdir -p $out/share/applications $out/share/icons/hicolor/256x256/apps
+    ln -s $LBRY_PATH/lbry.png $out/share/icons/hicolor/256x256/lbry.png
+    ln -s $LBRY_PATH/lbry.desktop $out/share/applications/lbry.desktop
+    substituteInPlace $LBRY_PATH/lbry.desktop \
+      --replace AppRun $out/bin/lbry
+
+    runHook postInstall
+  '';
+}
```
Comment (hjones2199) [inline pkgs/applications/video/lbry/default.nix:52]:
```suggestion
  meta = with lib; {
    homepage = "https://lbry.com/";
    description = "A graphical browser for the decentralized content marketplace provided by the LBRY protocol";
    license = licenses.mit;
    maintainers = with maintainers; [ enderger ];
    platforms = [ "x86_64-linux" ];
  };
}
```
You need a meta section, and you probably want to set yourself as maintainer.

**Review by hjones2199 — CHANGES_REQUESTED** [review]
Lacking meta section, and is extracting from the AppImage the best way to build this? It may unecessarily limit availability if upstream supports non x86_64 architectures.

File: pkgs/applications/video/lbry/default.nix (hunk 1 above)
Comment (enderger) [reply pkgs/applications/video/lbry/default.nix:52]: Oh, oops! Sorry about that, I had a previous version trying to build from source but I couldn't get electron-builder to work and decided to switch to appimage (tried for ~3 days with no success)

**Comment by enderger** [comment]
> Lacking meta section, and is extracting from the AppImage the best way to build this? It may unecessarily limit availability if upstream supports non x86_64 architectures.

I spent roughly 3 days trying to get electron-builder to work, but kept hitting roadblocks and ending up with an unwieldy solution. Also, packaging for OSX is not really viable, as Electron on MacOS uses a different file layout to Linux.

As for the meta issue, that is my fault, I didn't realize I hadn't ported that from the aforementioned from-source attempt.

File: pkgs/applications/video/lbry/default.nix (hunk 1 above)
Comment (enderger) [reply pkgs/applications/video/lbry/default.nix:52]: Also, I can confirm that upstream does not support non-x86_64 (with ARM locked out due to a need to build from source using electron-builder, something which I have failed to do after much effort)

**Comment by enderger** [comment]
Fixed the meta issue, didn't know how else to squash the commits (sorry if that was wrong, still a bit new to Git)

File: pkgs/applications/video/lbry/default.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    license = licenses.mit;
+    homepage = "https://lbry.com/";
+    downloadPage = "https://lbry.com/get/";
+    changelog = "https://github.com/lbryio/lbry-desktop/blob/master/CHANGELOG.md";
+    maintainers = with maintainers; [ enderger ];
+    platforms = [ "x86_64-linux" ];
+  };
+}
+
+
```
Comment (hjones2199) [inline pkgs/applications/video/lbry/default.nix:67]:
```suggestion

```
One superfluous line at end of file

**Review by hjones2199 — CHANGES_REQUESTED** [review]

**Comment by hjones2199** [comment]
> > Lacking meta section, and is extracting from the AppImage the best way to build this? It may unecessarily limit availability if upstream supports non x86_64 architectures.
> 
> I spent roughly 3 days trying to get electron-builder to work, but kept hitting roadblocks and ending up with an unwieldy solution. Also, packaging for OSX is not really viable, as Electron on MacOS uses a different file layout to Linux.
> 
> As for the meta issue, that is my fault, I didn't realize I hadn't ported that from the aforementioned from-source attempt.



> Fixed the meta issue, didn't know how else to squash the commits (sorry if that was wrong, still a bit new to Git)

You did that correctly, force-pushing a fixup into the original commit is the typical way to do that (at least in nixpkgs).

File: pkgs/applications/video/lbry/default.nix (hunk 2 above)
Comment (enderger) [reply pkgs/applications/video/lbry/default.nix:67]: Ah, looks like that's another on-save script I need to add to my vimrc. I'll fix that.

**Comment by hjones2199** [comment]
Result of `nixpkgs-review pr 119281` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lbry</li>
  </ul>
</details>

**Comment by enderger** [comment]
Fixed the whitespace issue.

**Review by hjones2199 — COMMENTED** [review]
While it's building locally with nixpkgs-review, ofborg bot is failing due to some issue with the xdg_utils dependency (which by the way is also not in your buildInputs or nativeBuildInputs).

I also have issues running the application once built. I'm guessing it has something to do with the way the extracted appimage is being wrapped. You might want to look into how other appimage-extracted packages use appimageTools.wrapType2. An example being [nuclear](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/nuclear/default.nix)

**Comment by enderger** [comment]
I'll see if it works better when added to buildInputs, perhaps that's not being added because of that.

**Comment by enderger** [comment]
As for the execution issue, I've built and run it several times without issues. Perhaps that is also linked to the buildinputs issue.

**Comment by enderger** [comment]
Hm, I’ll have to debug this in the morning, though I can’t seem to reproduce the execution problems on my machine. I’ll see if my NixOS main tower has the same problem (was developing this package off my school laptop running Fedora, so impurity may be leaking through).

File: pkgs/applications/video/lbry/default.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    ln -s $LBRY_PATH/lbry.png $out/share/icons/hicolor/256x256/lbry.png
+    ln -s $LBRY_PATH/lbry.desktop $out/share/applications/lbry.desktop
+    substituteInPlace $LBRY_PATH/lbry.desktop \
+      --replace AppRun $out/bin/lbry
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A browser and wallet for LBRY, the decentralized, user-controlled content marketplace.";
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lbry/default.nix:54]:
```suggestion
    description = "A browser and wallet for LBRY, the decentralized, user-controlled content marketplace";
```

File: pkgs/applications/video/lbry/default.nix [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    longDescription = ''
+      The LBRY app is a graphical browser for the decentralized content marketplace provided by the LBRY protocol. It is essentially the lbry daemon bundled with a UI using Electron.
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lbry/default.nix:56]:
```suggestion
      The LBRY app is a graphical browser for the decentralized content marketplace provided by the LBRY protocol.
      It is essentially the lbry daemon bundled with a UI using Electron.
```

File: pkgs/applications/video/lbry/default.nix [hunk 5]
```diff
@@ -0,0 +1,65 @@
+{ lib
+, fetchurl
+, stdenv
+, appimageTools
+, ffmpeg
+, gtk3
+, makeWrapper
+, xdg_utils
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lbry/default.nix:8]:
```suggestion
, xdg-utils
```

File: pkgs/applications/video/lbry/default.nix [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  src = appimageTools.extract {
+    name = pname;
+    src = fetchurl {
+      url = "https://github.com/lbryio/lbry-desktop/releases/download/v${version}/LBRY_${version}.AppImage";
+      # Gotten from latest-linux.yml
+      sha512 = "br6HvVRz+ybmAhmQh3vOC5wgLmOCVrGHDn59ueWk6rFoKOCbm8WdmdadOZvHeN1ld2nlvPzEy+KXMOEfF1LeQg==";
+    };
+  };
+
+  buildInputs = [ ffmpeg gtk3 xdg_utils ];
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lbry/default.nix:25]:
```suggestion
  buildInputs = [ ffmpeg gtk3 xdg-utils ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/video/lbry/default.nix (hunk 5 above)
Comment (enderger) [reply pkgs/applications/video/lbry/default.nix:8]: Ah, xdg_utils is now an alias. Makes sense.

Comment (enderger) [reply pkgs/applications/video/lbry/default.nix:8]: BTW, while importing these changes I found another instance of xdg_utils, so self-importing makes sense here.

**Comment by enderger** [comment]
Alright, I've fixed the issues (should probably have Grepped the `aliases.nix` file for XDG while debugging the Ofborg issue, sorry about that).

**Comment by hjones2199** [comment]
Result of `nixpkgs-review pr 119281` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lbry</li>
  </ul>
</details>

**Comment by hjones2199** [comment]
I think theres still some impurity, perhaps hard-coded paths in the executable? running it gives me:
`./results/lbry/bin/lbry: line 4: /nix/store/vlcdkjbgm8v6k0w2v39f88y0r2vk8i4w-lbry-desktop-0.50.2/lib/lbry-desktop/lbry: No such file or directory`

**Comment by enderger** [comment]
> I think theres still some impurity, perhaps hard-coded paths in the executable? running it on NixOS gives me:
> `./results/lbry/bin/lbry: line 4: /nix/store/vlcdkjbgm8v6k0w2v39f88y0r2vk8i4w-lbry-desktop-0.50.2/lib/lbry-desktop/lbry: No such file or directory`

Odd, can you try `ls result/lib/lbry-desktop/`?

**Comment by enderger** [comment]
Ah, I think it's the app icon, as that looks like a symlink in the AppImage. I'll see about getting that fixed.

**Comment by enderger** [comment]
Hey, would you happen to know how I would go about debigging this issue? I have no idea.

**Comment by enderger** [comment]
Ok, it looks like the issue stemmed from environment. I've moved over to the wrapAppImage helper and gotten it working (in a more elegant way, to boot).

**Comment by hjones2199** [comment]
Result of `nixpkgs-review pr 119281` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lbry</li>
  </ul>
</details>

**Review by hjones2199 — APPROVED** [review]

**Comment by enderger** [comment]
@SuperSandro2000 - When you have a moment, could you re-review this PR? I've gotten it building on NixOS.

**Comment by enderger** [comment]
Thanks!

**Pull request merged**